### PR TITLE
 [Typing][B-97] Add type annotations for `python/paddle/distributed/collective.py`

### PR DESCRIPTION
**Pull request by enkilee**
### PR Category

User Experience

### PR Types

Improvements

### Description

- https://github.com/PaddlePaddle/Paddle/issues/65008
B97

### CODE REVIEW

File: python/paddle/distributed/collective.py [hunk 1]
```diff
@@ -184,11 +193,11 @@ def _set_custom_gid(gid):
 
 
 def new_group(
-    ranks=None,
-    backend=None,
-    timeout=_default_timeout,
-    nccl_comm_init_option=0,
-):
+    ranks: list | None = None,
```
Comment (megemini) [inline python/paddle/distributed/collective.py]:
```suggestion
    ranks: list[int] | None = None,
```

File: python/paddle/distributed/collective.py [hunk 2]
```diff
@@ -184,11 +193,11 @@ def _set_custom_gid(gid):
...
+    backend: str | None = None,
```
Comment (megemini) [inline python/paddle/distributed/collective.py]: 这里貌似只能是 `nccl` ？

File: python/paddle/distributed/collective.py (hunk 1 above)
Comment (enkilee) [reply python/paddle/distributed/collective.py]: 收到

File: python/paddle/distributed/collective.py (hunk 2 above)
Comment (enkilee) [reply python/paddle/distributed/collective.py]: 收到

**Review by SigureMo — APPROVED** [review]
LGTMeow

**Pull request merged**